### PR TITLE
[cxxmodules] Preload all modules.

### PR DESCRIPTION
**Pull request by vgvassilev**
This does not seem to affect the startup performance of ROOT.

### CODE REVIEW

**Comment by vgvassilev** [comment]
@phsft-bot build just on slc6/clang_gcc62 with flags -Dcxxmodules=On -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@phsft-bot build just on centos7/gcc62 with flags -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

File: core/metacling/src/TCling.cxx [hunk 1]
```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
...
+   // do better: use the global module index to avoid loading all modules and load only the ones that contain the
+   // looked up identifier.
+   for (auto M : TopLevelModulesToLoad) {
+      assert(M);
+      assert(M->IsSystem || M->IsMissingRequirement);
+      // Skip Core and RIO: we alread loaded them.
+      if (M == CoreM || M == RIOM)
+        continue;
+      TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, M);
+      for (clang::Module *subModule : M->submodules())
```
Comment (Teemperor) [inline core/metacling/src/TCling.cxx]: Modules and submodules form a tree, so just going over the submodules isn't enough as you miss sub-submodules that way. Easy way to do this is:
```C++
   llvm::SetVector<clang::Module *> modules;
   modules.insert(module);
   for (size_t i = 0; i < modules.size(); ++i) {
      clang::Module *M = modules[i];
      for (clang::Module *subModule : M->submodules()) modules.insert(subModule);
   }
```

File: core/metacling/src/TCling.cxx [hunk 2]
```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
...
+   clang::Sema &TheSema = CI.getSema();
+   // createImplicitModuleImportForErrorRecovery creates decls.
+   cling::Interpreter::PushTransactionRAII RAII(&interp);
+   std::vector<clang::Module*> TopLevelModulesToLoad;
+   for (auto I = moduleMap.module_begin(), E = moduleMap.module_end(); I != E; ++I) {
+     clang::Module *M = I->second;
+     assert(M);
+     if (M->IsSystem && !M->IsMissingRequirement) {
+       TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, M);
+       for (clang::Module *subModule : M->submodules())
```
Comment (Teemperor) [inline core/metacling/src/TCling.cxx]: Same problem as below with the submodules.

File: core/metacling/src/TCling.cxx [hunk 3]
```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
...
-   if (result != cling::Interpreter::CompilationResult::kSuccess) {
-      Error("TCling::LoadCoreModules", "Couldn't parse core headers: %s\n", declarations.str().c_str());
+   // Loading all other modules in rootcling would mean injecting extra (needless) dependencies to the pcms.
+   bool fromRootCling = dlsym(RTLD_DEFAULT, "usedToIdentifyRootClingByDlSym");
+   if (fromRootCling)
+     return;
+
+   // FIXME: The performance of loading all modules at start up is comparable to the non-modules one. However we can
+   // do better: use the global module index to avoid loading all modules and load only the ones that contain the
+   // looked up identifier.
```
Comment (Teemperor) [inline core/metacling/src/TCling.cxx]: I think we already asked Richard about that and he said it's only storing identifiers and therefor only useful in C (not C++).

File: core/metacling/src/TCling.cxx [hunk 4]
```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
...
+   for (auto M : TopLevelModulesToLoad) {
+      assert(M);
+      assert(M->IsSystem || M->IsMissingRequirement);
+      // Skip Core and RIO: we alread loaded them.
+      if (M == CoreM || M == RIOM)
```
Comment (Teemperor) [inline core/metacling/src/TCling.cxx]: More generic:

```C++
if (TheSema.isModuleVisible(M))
```

File: core/metacling/src/TCling.cxx [hunk 5]
```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
...
+     if (M->IsSystem && !M->IsMissingRequirement) {
+       TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, M);
+       for (clang::Module *subModule : M->submodules())
+         TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, subModule);
+     }
+     else if (M->IsSystem && !M->IsMissingRequirement)
+       TopLevelModulesToLoad.push_back(M);
+   }
+   // Those two modules are required to be preloaded by ROOT. If they are not around we have a big problem.
+   clang::Module *CoreM = moduleMap.findModule("Core");
```
Comment (Teemperor) [inline core/metacling/src/TCling.cxx]: Why not recursively loading the submodules here? Does it work? (never tested that).

**Comment by Teemperor** [comment]
I think we want in general a function that makes a module with submodules available. We also could/should use that in `RegisterModule` to not invoke the parser there (which probably fixes some other problems too, thanks to @Axel-Naumann for this idea).

**Comment by vgvassilev** [comment]
@phsft-bot build just on centos7/clang_gcc62 with flags -Dcxxmodules=On -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Comment by vgvassilev** [comment]
@phsft-bot build just on slc6/clang_gcc62 with flags -Dcxxmodules=On -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Pull request closed without merging**